app/frontend/wallet/cardano-ledger-crypto-provider.js

```diff
@@ -0,0 +1,128 @@
+const TransportU2F = require('@ledgerhq/hw-transport-u2f').default // for browser
+const Ada = require('ledgerhq/hw-app-ada').default // temporary hack, should be @ledgerhq
+const CachedDeriveXpubFactory = require('./helpers/CachedDeriveXpubFactory')
+const cbor = require('borc')
+const {TxWitness, SignedTransactionStructured} = require('./transaction')
+
+const CardanoLedgerCryptoProvider = async (ADALITE_CONFIG, walletState) => {
+  const transport = await TransportU2F.create()
+  const ada = new Ada(transport)
+
+  const state = Object.assign(walletState, {
+    rootHdPassphrase: null,
+    derivedAddresses: {},
+  })
+
+  if (state.derivationScheme.type !== 'v2') {
+    throw new Error(`Unsupported derivation scheme: ${state.derivationScheme.type}`)
+  }
+
+  const deriveXpub = CachedDeriveXpubFactory(state.derivationScheme, async (absDerivationPath) => {
+    try {
+      const response = await ada.getExtendedPublicKey(absDerivationPath)
+      return Buffer.from(response.publicKeyHex.concat(response.chainCodeHex), 'hex')
+    } catch (err) {
+      throw new Error('public key retrieval from Ledger failed')
+    }
+  })
+
+  function deriveHdNode(childIndex) {
+    throw new Error('This operation is not supported on LedgerCryptoProvider!')
+  }
+
+  function sign(message, absDerivationPath) {
+    throw new Error('Not supported')
+  }
+
+  async function displayAddressForPath(absDerivationPath) {
+    try {
+      await ada.showAddress(absDerivationPath)
+    } catch (err) {
+      throw new Error('Ledger operation failed!')
+    }
+  }
+
+  function getWalletSecret() {
+    throw new Error('Unsupported operation!')
+  }
+
+  function prepareInput(input, addressToAbsPathMapper, txDataHex) {
+    const data = {
+      txDataHex,
+      outputIndex: input.outputIndex,
+      path: addressToAbsPathMapper(input.utxo.address),
+    }
+
+    return data
+  }
+
+  function prepareOutput(output, addressToAbsPathMapper) {
+    const data = {
+      amountStr: `${output.coins}`,
+    }
+
+    if (output.isChange) {
+      data.path = addressToAbsPathMapper(output.address)
+    } else {
+      data.address58 = output.address
+    }
+
+    return data
+  }
+
+  async function prepareWitnesses(witnesses) {
+    const data = []
+    for (const witness of witnesses) {
```

Comment:
I would rewrite this using Array.map function:
```
   witnesses.map((witness) => {
      ...
   })
```
It is a good practice to replace "for" cycles with map, reduce and similar functions in js, because it improves readability - it makes clearer the purpose of the iteration and it also eliminates the need to have a "data" array aside - it directly returns it, so you can assign the result to data